Data/Graph/Inductive/Query/Cycles.hs

```diff
@@ -0,0 +1,280 @@
+{- |
+   Module      : Data.Graph.Inductive.Query.Cycles
+   Description : Algorithms for finding all cycles.
+   Copyright   : (c) Gabriel Hjort Blindell 2017
+   License     : 2-Clause BSD
+   Maintainer  : gabriel.hjort.blindell@gmail.com
+
+   Defines algorithms that find all cycles in a given graph.
+ -}
+module Data.Graph.Inductive.Query.Cycles
+  ( cyclesIn
+  , cyclesIn'
+  , strongComponentsOf
+  , uniqueCycles
+  , uniqueCycles'
+  )
+where
+
+import Data.Graph.Inductive.Graph
+
+import Data.List
+  ( (\\)
+  , delete
+  , tails
+  )
+import Data.Maybe
+  ( fromJust )
+import Control.Monad
+  ( ap )
+import qualified Data.Map as M
+
+
+
+-- | Obtain the labels for a list of 'Node's.
+--   It is assumed that each 'Node' is indeed present in the given graph.
+addLabels :: (Graph g) => g a b -> [Node] -> [LNode a]
+addLabels gr = map (ap (,) (fromJust . lab gr))
+
+twoCycle      :: (Graph g) => g a b -> Node -> [Node]
+twoCycle gr n = filter (elem n . suc gr) (delete n $ suc gr n)
+
+-- | Determines if the list of nodes represents a regular subgraph.
+isRegular :: (Graph g) => g a b -> [Node] -> Bool
+isRegular g ns = all allTwoCycle split
+    where
+      -- Node + Rest of list
+      split = zip ns tns'
+      tns' = tail $ tails ns
+      allTwoCycle (n,rs) = null $ rs \\ twoCycle g n
+
+-- | Contains the necessary data structures used by 'strongComponentsOf'.
+data SCCState g a b
+  = SCCState
+      { sccComponents :: [g a b]
+        -- ^ The components found so far.
+      , sccCurrentIndex :: Int
+        -- ^ The current index.
+      , sccStack :: [Node]
+        -- ^ The node stack.
+      , sccNodeInfo :: M.Map Node (Bool, Int, Int)
+        -- ^ Node information as a tuple (whether the node is on the stack, its
+        -- index, and its low link).
+      , sccGraph :: g a b
+        -- ^ The input graph.
+      }
+
+-- | Find all strongly connected components of a graph. Implements Tarjan's
+-- algorithm. Returned list is sorted in topological order.
+strongComponentsOf :: (DynGraph g) => g a b -> [g a b]
+strongComponentsOf g =
+  sccComponents $
+  foldr ( \n st ->
+          let (_, i, _) = sccNodeInfo st M.! n
+          in if i < 0 then findSCCFor n st else st
+        )
+        (mkInitSCCState g)
+        (nodes g)
+
+findSCCFor :: (DynGraph g) => Node -> SCCState g a b -> SCCState g a b
+findSCCFor n st0 =
+  let i = sccCurrentIndex st0
+      st1 = st0 { sccCurrentIndex = i + 1
+                , sccStack = (n:sccStack st0)
+                , sccNodeInfo = M.insert n (True, i, i) (sccNodeInfo st0)
+                }
+      g = sccGraph st1
+      st2 = foldr ( \m st ->
```

Comment:
This is large enough that I think a new function - even if it's in a where block - is warranted for documentation, etc.

That way, you can also use guards rather than nested if/else.